### PR TITLE
Implement Egress NPC

### LINKED ISSUE #2624: Support k8s egress policies
_From @awh on September 21, 2016 9:6_

The k8s network SIG is developing an egress policy specification which we should support.

_Copied from original issue: weaveworks/weave-npc#9_

### PR DESCRIPTION
**Pull request by brb**
This PR implements Kubernetes Egress Network Policy (defined in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#networkpolicyspec-v1-networking-k8s-io).

Some important changes are mentioned/explained below.

### Rules

Each "weave" packet (=sent over the `weave` bridge) is first checked against egress and then against ingress rules. 

For Egress NPC we introduce the following rules:

```
    -A WEAVE-NPC-EGRESS -m state --state RELATED,ESTABLISHED -j RETURN
    -A WEAVE-NPC-EGRESS -m set ! --match-set weave-local-pods src -j RETURN
    -A WEAVE-NPC-EGRESS -d 224.0.0.0/4 -j RETURN
    -A WEAVE-NPC-EGRESS -m state --state NEW -j WEAVE-NPC-EGRESS-DEFAULT
    -A WEAVE-NPC-EGRESS -m state --state NEW -m mark ! --mark 0x40000/0x40000 -j WEAVE-NPC-EGRESS-CUSTOM
    -A WEAVE-NPC-EGRESS -m state --state NEW -m mark ! --mark 0x40000/0x40000 -j NFLOG --nflog-group 86
    -A WEAVE-NPC-EGRESS -m mark ! --mark 0x40000/0x40000 -j DROP

    -A WEAVE-NPC-EGRESS-CUSTOM <rulespec> -j MARK --set-xmark 0x40000/0x40000
    -A WEAVE-NPC-EGRESS-CUSTOM <rulespec> -j RETURN

    -A WEAVE-NPC-EGRESS-DEFAULT <rulespec> -j MARK --set-xmark 0x40000/0x40000
    -A WEAVE-NPC-EGRESS-DEFAULT <rulespec> -j RETURN
```

When a rule which matches an egress packet is found, we mark the packet instead of ACCEPT'ing. This is because the same packet need to traverse the `WEAVE-NPC` chain (Ingress), so we cannot just accept the packet.

Also, in opposite to the ingress rules, we move the logging and dropping into the `WEAVE-NPC-EGRESS` chain. We ensure that the rules are not flushed upon a restart of weave-npc, so that no time window is created when all egress traffic is allowed.

### Traffic Steering

To steer the traffic for the egress NPC processing we create two rules:

- `-A INPUT -i weave -j WEAVE-NPC-EGRESS`: applies to packets destined to a local node.
- `-A FORWARD -i weave -j WEAVE-NPC-EGRESS`: the rest. It can match an ingress packet.

### ipBlock

The Egress NPC supports the ipBlock selector. Some of its implementation has been cherry-picked from https://github.com/weaveworks/weave/pull/3233. Note, that we still do not support the selector in ingress!

### Renaming dstSelector to targetSelector
 
We have renamed "dstSelector" to "targetSelector" because "dstSelector" becomes ambiguous in the context of egress. Just to recall, we call a selector a "targetSelector" if it selects pods to which a network policy applies (=`.NetworkPolicySpec.podSelector`).

### Del --use-legacy-netpol

We are no longer supporting the legacy network policies.

Fix #2624

### CODE REVIEW

**Comment by brb** [comment]
PTAL.

File: npc/rule.go [hunk 1]
```diff
@@ -32,13 +31,13 @@ func newRuleSpec(policyType policyType, proto *string, srcHost, dstHost ruleHost
 		args = append(args, "-p", *proto)
 	}
 	srcComment := "anywhere"
-	if srcHost != nil && !srcHost.isNil() {
+	if srcHost != nil && !reflect.ValueOf(srcHost).IsNil() {
```
Comment (bboreham) [inline npc/rule.go]: what's going on here?  how does `srcHost` git to be nil-valued but not `nil` ?

Comment (brb) [reply npc/rule.go]: Go :man_shrugging: https://play.golang.org/p/bCC2ztY1NBh

Comment (bboreham) [reply npc/rule.go]: ok, it is `nil`, but of a different type to the one you asked about.

**Comment by hswong3i** [comment]
Could we make it commit so a 2.4.0 release?

File: npc/analyser.go [hunk 2]
```diff
@@ -4,173 +4,158 @@ import (
...
 				}
+			} else {
+				for _, peer := range egressRule.To {
+					var dstSelector *selectorSpec
+					var dstRuleHost ruleHost
+
+					// NetworkPolicyPeer describes a peer to allow traffic to.
+					// Exactly one of its fields must be specified.
+					if peer.PodSelector != nil {
+						dstSelector, err = newSelectorSpec(peer.PodSelector, nil, ns.name, ipset.HashIP)
```
Comment (murali-reddy) [inline npc/analyser.go]: May be a separate PR is just fine, but just wanted to call out the semantic changes in network policy.
https://github.com/kubernetes/kubernetes/issues/58637
Both pod selector and namespace selector can be specified together.

File: net/bridge.go [hunk 3]
```diff
@@ -457,14 +458,27 @@ func configureIPTables(config *BridgeConfig, ips ipset.Interface) error {
...
 
 	if config.NPC {
-		// Steer traffic via the NPC
-		_ = ipt.NewChain("filter", "WEAVE-NPC") // ignore error because it might already exist
-		// If WEAVE-NPC chain doesn't exist then creating a rule in the chain will fail
+		// Steer traffic via the NPC.
+
+		// Ignore error because chains might already exist
+		_ = ipt.NewChain("filter", npc.MainChain)
+		_ = ipt.NewChain("filter", npc.EgressChain)
```
Comment (murali-reddy) [inline net/bridge.go]: May be check if chain exists by `ListChains` of the `filter` table? Just ignoring the error might mask any real error.

File: npc/analyser.go (hunk 2 above)
Comment (brb) [reply npc/analyser.go]: Thanks for noting, I'm going to address this in a separate PR (https://github.com/weaveworks/weave/issues/3312).

File: net/bridge.go (hunk 3 above)
Comment (brb) [reply net/bridge.go]: Agreed!

File: net/bridge.go [hunk 4]
```diff
@@ -457,14 +458,27 @@ func configureIPTables(config *BridgeConfig, ips ipset.Interface) error {
...
+		if err = ipt.AppendUnique("filter", "INPUT", "-i", config.WeaveBridgeName, "-j", npc.EgressChain); err != nil {
+			return err
+		}
 		fwdRules = append(fwdRules,
 			[][]string{
+				// Might include ingress traffic which is fine as long as we do not
+				// ACCEPT in WEAVE-NPC-EGRESS chain
+				{"-i", config.WeaveBridgeName,
+					"-m", "comment", "--comment", "NOTE: this must go before '-j KUBE-FORWARD'",
+					"-j", npc.EgressChain},
```
[the remark this answers is not in the document]
Comment (brb) [reply net/bridge.go]: I don't see how we could do it as in the case of egress there are two entry points - `-A INPUT -i weave -j WEAVE-NPC-EGRESS` and `-A FORWARD -i weave -j WEAVE-NPC-EGRESS`, and in the case of ingress only one - `-A FORWARD -o weave -j WEAVE-NPC`. Also, the direction of the `weave` interface in the `FORWARD` chain differs.

File: prog/weave-npc/main.go [hunk 5]
```diff
@@ -112,11 +137,93 @@ func createBaseRules(ipt *iptables.IPTables, ips ipset.Interface) error {
...
-	return ipt.Append(npc.TableFilter, npc.MainChain,
-		"-m", "set", "!", "--match-set", npc.LocalIpset, "dst", "-j", "ACCEPT")
+
+	if err := ipt.Append(npc.TableFilter, npc.EgressMarkChain,
+		"-j", "MARK", "--set-xmark", npc.EgressMark); err != nil {
+		return err
+	}
+
+	ruleSpecs := [][]string{
+		{"-m", "state", "--state", "RELATED,ESTABLISHED", "-j", "RETURN"},
```
[the remark this answers is not in the document]
Comment (brb) [reply prog/weave-npc/main.go]: Good suggestion! I just need to check whether we do not miss a traversal of any kube-proxy rule due to an early ACCEPT.

File: prog/weave-npc/main.go [hunk 6]
```diff
@@ -112,11 +137,93 @@ func createBaseRules(ipt *iptables.IPTables, ips ipset.Interface) error {
...
+		{"-m", "state", "--state", "NEW", "-m", "set", "!", "--match-set", npc.LocalIpset, "src", "-j", "RETURN"},
```
[the remark this answers is not in the document]
Comment (brb) [reply prog/weave-npc/main.go]: Unfortunately not - the fact that incoming interface is `weave` does not guarantee that a packet originated on a local node.

File: npc/namespace.go [hunk 7]
```diff
@@ -369,73 +393,89 @@ func (ns *ns) deleteNetworkPolicy(obj interface{}) error {
...
+func bypassRules(namespace string, ingress, egress ipset.Name) map[string][][]string {
+	return map[string][][]string{
+		DefaultChain: {
+			{"-m", "set", "--match-set", string(ingress), "dst", "-j", "ACCEPT",
+				"-m", "comment", "--comment", "DefaultAllow ingress isolation for namespace: " + namespace},
+		},
+		EgressDefaultChain: {
+			{"-m", "set", "--match-set", string(egress), "src", "-j", EgressMarkChain,
+				"-m", "comment", "--comment", "DefaultAllow egress isolation for namespace: " + namespace},
+			{"-m", "set", "--match-set", string(egress), "src", "-j", "RETURN",
```
[the remark this answers is not in the document]
Comment (brb) [reply npc/namespace.go]: > I believe currently WEAVE-NPC-EGRESS chain never does ACCEPT packet.

Yep, we don't ACCEPT.

> IMO we should accept the packets here explicitly.

I don't see how as in some cases we need to traverse both sets of rules - egress and ingress (e.g. pod to pod running on the same node).

[the remark this answers is not in the document]
Comment (murali-reddy) [reply npc/namespace.go]: Ag, got it. May be a comment somewhere to explain the flow (across the chains for both ingress and egress) would be helpful.

File: prog/weave-npc/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (murali-reddy) [reply prog/weave-npc/main.go]: If the packet  has arrived from the `weave` bridge interface, is it not sufficient condition to say packet is from one the pods running locally? 
Is `LocalIpset` inclusive of pods running in host network?

[the remark this answers is not in the document]
Comment (brb) [reply prog/weave-npc/main.go]: > is it not sufficient condition to say packet is from one the pods running locally?

Unfortunately noup, because we connect a datapath to the `weave` bridge via `vethwe-bridge` link, so the bridge receives weave traffic from remote hosts as well.

> Is LocalIpset inclusive of pods running in host network?

It just includes a host local pods, but I think we can get rid of the ipset in favor of https://github.com/weaveworks/weave/issues/3344

File: npc/namespace.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (brb) [reply npc/namespace.go]: Good idea! But where would you suggest to put this comment? I already have it in this PR description (see above) and also in the commit message: https://github.com/weaveworks/weave/pull/3313/commits/af622dc0ef81aabf1c40a29ac5def0fd5c0153c2

[the remark this answers is not in the document]
Comment (murali-reddy) [reply npc/namespace.go]: Understand you already added it in the PR and commit. Just suggesting some comment in the code may be in `prog/weave-npc/main.go` around `createBaseRules` or wherever you feel appropriate. Basically to give some sense of how ingress and egress network policies are enforced for someone reading the code.

File: net/bridge.go (hunk 3 above)
Comment (brb) [reply net/bridge.go]: I've introduced `ensureChains` function which does what you suggest.

File: prog/weave-npc/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (brb) [reply prog/weave-npc/main.go]: I've changed the target to `ACCEPT`.

File: npc/namespace.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (brb) [reply npc/namespace.go]: I've put a comment to `createBaseRules`.

**Comment by murali-reddy** [comment]
@brb FWIW, changes look good to me for merge
